File: swift/Tasks/Services/DittoManager.swift
import DittoSwift
import Foundation

/// Owner of the Ditto object
class DittoManager: ObservableObject {
    let ditto: Ditto
    static let shared = DittoManager()
    
    var syncScopes = [
      "local_mesh_only": "SmallPeersOnly"
    ]

    func getDeviceName() -> String {
        return CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "Unset"
    }

    private init() {
        // https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
        ditto = Ditto(
            identity: .onlinePlayground(
                appID: Env.DITTO_APP_ID,
                token: Env.DITTO_PLAYGROUND_TOKEN,
                // This is required to be set to false to use the correct URLs
                // This only disables cloud sync when the webSocketURL is not set explicitly
                enableDittoCloudSync: false,
                customAuthURL: URL(string: Env.DITTO_AUTH_URL)
            )
        )

        ditto.deviceName = getDeviceName()
        // Set the Ditto Websocket URL
        ditto.updateTransportConfig { transportConfig in
            //transportConfig.connect.webSocketURLs.insert(Env.DITTO_WEBSOCKET_URL)
            transportConfig.peerToPeer.bluetoothLE.isEnabled = true
            transportConfig.peerToPeer.awdl.isEnabled = true
            transportConfig.peerToPeer.lan.isEnabled = true
        }

        // disable sync with v3 peers, required for DQL
        do {
            try ditto.disableSyncWithV3()
            Task {
                try await ditto.store.execute(query: "ALTER SYSTEM SET DQL_STRICT_MODE = false")
                try await ditto.store.execute(query: "ALTER SYSTEM SET rotating_log_file_max_size_mb  =\(Env.DITTO_LOG_SIZE)")
                
                try await ditto.store.execute(
                    query: "ALTER SYSTEM SET USER_COLLECTION_SYNC_SCOPES = :syncScopes",
                    arguments: ["syncScopes": syncScopes]
                )
            }
        } catch let error {
            print(
                "DittoManger - ERROR: disableSyncWithV3() failed with error \"\(error)\""
            )
        }

        let isPreview: Bool =
            ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
        if !isPreview {
            DittoLogger.minimumLogLevel = .debug
        }
    }
}

File: android-kotlin/QuickStartTasks/app/src/main/java/live/ditto/quickstart/tasks/DittoManager.kt
package live.ditto.quickstart.tasks

import android.content.Context
import android.util.Log
import kotlinx.coroutines.flow.Flow
import live.ditto.Ditto
import live.ditto.DittoError
import live.ditto.DittoIdentity
import live.ditto.DittoLogger
import live.ditto.DittoLogLevel
import live.ditto.DittoQueryResult
import live.ditto.DittoSyncSubscription
import live.ditto.android.DefaultAndroidDittoDependencies
import live.ditto.transports.DittoSyncPermissions
import javax.inject.Singleton

@Singleton
class DittoManager(
    private val context: Context,
) {
    companion object {
        private const val TAG: String = "DittoManager"
    }

    private lateinit var ditto: Ditto
    private val currentSubscriptions: MutableMap<DittoCollectionSubscription, DittoSyncSubscription> = mutableMapOf()

    val syncScopes = mapOf(
        "local_mesh_only" to "SmallPeersOnly"
    )

    suspend fun initDitto() {
        try {
            DittoLogger.minimumLogLevel = DittoLogLevel.DEBUG

            val androidDependencies = DefaultAndroidDittoDependencies(context)

            val identity = DittoIdentity.OnlinePlayground(
                dependencies = androidDependencies,
                appId = BuildConfig.DITTO_APP_ID,
                token = BuildConfig.DITTO_PLAYGROUND_TOKEN,
                customAuthUrl = BuildConfig.DITTO_AUTH_URL,
                enableDittoCloudSync = false
            )

            ditto = Ditto(
                dependencies = androidDependencies,
                identity = identity
            ).apply {
                disableSyncWithV3()

                updateTransportConfig { transportConfig ->
                    //transportConfig.connect.websocketUrls.add(BuildConfig.DITTO_WEBSOCKET_URL)
                }

                store.execute(
                    "ALTER SYSTEM SET USER_COLLECTION_SYNC_SCOPES = :syncScopes",
                    mapOf("syncScopes" to syncScopes)
                )
                store.execute("ALTER SYSTEM SET rotating_log_file_max_size_mb = ${BuildConfig.DITTO_LOG_SIZE}")
                store.execute("ALTER SYSTEM SET DQL_STRICT_MODE = false")
                startSync()
            }

        } catch (e: DittoError) {
            Log.e(TAG, e.message ?: "Failed to initialize Ditto")
            throw e
        }
    }

    fun requireDitto(): Ditto {
        if (::ditto.isInitialized) {
            return ditto
        } else {
            throw DittoNotInitializedException()
        }
    }

    fun isDittoInitialized(): Boolean {
        return ::ditto.isInitialized
    }

    fun missingPermissions(): Array<String> {
        return DittoSyncPermissions(context = context).missingPermissions()
    }

    suspend fun executeQuery(query: String, args: Map<String, Any?>?): Result<DittoQueryResult> {
        return try {
            if (::ditto.isInitialized) {
                Result.success(ditto.store.execute(query, args))
            } else {
                Result.failure(DittoNotInitializedException())
            }
        } catch (e: Exception) {
            Log.e(TAG, "Query execution failed: ${e.message}", e)
            Result.failure(e)
        }
    }

    fun liveQueryAsFlow(subscriptionQuery: String, args: Map<String, Any?>): Flow<DittoQueryResult> {
        return ditto.store.registerObserverAsFlow(
            query = subscriptionQuery,
            params = args
        )
    }

    fun registerSubscription(dittoCollectionSubscription: DittoCollectionSubscription) {
        val dittoSyncSubscription = ditto.sync.registerSubscription(
            query = dittoCollectionSubscription.subscriptionQuery,
            arguments = dittoCollectionSubscription.subscriptionQueryArgs
        )
        currentSubscriptions[dittoCollectionSubscription] = dittoSyncSubscription
    }

    fun startSync() {
        if (::ditto.isInitialized && !ditto.isSyncActive) {
            try {
                ditto.startSync()
            } catch (e: DittoError) {
                Log.e(TAG, "Unable to start sync", e)
                throw e
            }
        }
    }

    fun stopSync() {
        if (::ditto.isInitialized && ditto.isSyncActive) {
            try {
                ditto.stopSync()
            } catch (e: DittoError) {
                Log.e(TAG, "Unable to stop sync", e)
                throw e
            }
        }
    }

    fun isSyncActive(): Boolean {
        return if (::ditto.isInitialized) {
            ditto.isSyncActive
        } else {
            false
        }
    }

    fun toggleBluetoothLE() {
        if (::ditto.isInitialized) {
            ditto.updateTransportConfig { config ->
                config.peerToPeer.bluetoothLe.enabled = !config.peerToPeer.bluetoothLe.enabled
            }
        }
    }

    fun toggleLAN() {
        if (::ditto.isInitialized) {
            ditto.updateTransportConfig { config ->
                config.peerToPeer.lan.enabled = !config.peerToPeer.lan.enabled
            }
        }
    }

    fun toggleWifiAware() {
        if (::ditto.isInitialized) {
            ditto.updateTransportConfig { config ->
                config.peerToPeer.wifiAware.enabled = !config.peerToPeer.wifiAware.enabled
            }
        }
    }
}

class DittoNotInitializedException : IllegalStateException("Ditto has not been initialized. Call initDitto() first.")